engine/src/main/java/org/terasology/logic/health/HealthSystem.java

```diff
@@ -63,15 +63,25 @@ public void update(float delta) {
             }
 
             int healAmount = 0;
-            while (time.getGameTimeInMs() >= health.nextRegenTick) {
-                healAmount++;
-                health.nextRegenTick = health.nextRegenTick + (long) (1000 / health.regenRate);
-            }
+            healAmount = regenerateHealth(health, healAmount);
 
-            if (healAmount > 0) {
-                checkHeal(entity, healAmount, entity, health);
-                entity.saveComponent(health);
-            }
+            checkHealed(entity, health, healAmount);
+        }
+    }
+
+    private int regenerateHealth(HealthComponent health, int healAmount) {
+        int newHeal = healAmount;
+        while (time.getGameTimeInMs() >= health.nextRegenTick) {
+            newHeal++;
+            health.nextRegenTick = health.nextRegenTick + (long) (1000 / health.regenRate);
+        }
+        return newHeal;
+    }
+
+    private void checkHealed(EntityRef entity, HealthComponent health, int healAmount) {
```

Comment:
IIRC the distinction is between *any* heal and *fully* healed, so you can react to them differently? That's a vague memory tho.